### PR TITLE
Add additional Swapped instances

### PR DESCRIPTION
**Pull request by emilypi**
See: https://github.com/ekmett/lens/issues/753

@treeowl if you know of any other bifunctors with admissible `Swapped` instances, lets throw them on the pile!

### CODE REVIEW

**Comment by treeowl** [comment]
Perhaps add the sums and products from `GHC.Generics`?

**Comment by emilypi** [comment]
Ah righto i did miss those

File: src/Control/Lens/Iso.hs [hunk 1]
```diff
@@ -409,6 +412,47 @@ instance Swapped (,) where
...
+instance (Swapped p, Swapped q) => Swapped (Sum p q) where
+  swapped = iso f f
+    where
+      f (L2 x) = L2 (x ^. swapped)
+      f (R2 x) = R2 (x ^. swapped)
+
+instance (Functor f, Swapped p) => Swapped (Tannen f p) where
+  swapped = iso f f
+    where
+      f (Tannen x) = Tannen $ fmap (^. swapped) x
```
Comment (phadej) [inline src/Control/Lens/Iso.hs]: - How about `instance Swapped p => Swapped (Flip p)`?
- `(f ~ g, Functor f, Swapped p) => Swapped (Biff p f g)` is a bit on the edge

Comment (emilypi) [reply src/Control/Lens/Iso.hs]: Thanks @phadej! I was hoping to find a more satisfying `Biff` instance last night, but i suppose `f ~ g` is the best we get huh? 😄

Comment (phadej) [reply src/Control/Lens/Iso.hs]: Well, maybe `Coercible` bla blah (but that would be GHC >= 7.8), or some `Generic1` `Rep1` are isomorphic blah blah. I'm fine with just leaving it out too.

Comment (emilypi) [reply src/Control/Lens/Iso.hs]: I think we're good on instances now with the latest push. It turns out `:+:` and `:*:` are functors in `GHC.Generics`. I hadn't noticed before! very odd.

Comment (phadej) [reply src/Control/Lens/Iso.hs]: they are essentially `Product f g a` not `Product p q a b`

**Comment by treeowl** [comment]
Oh, sorry, I misread which `Product` and `Sum` those were. You now have a redundant `GHC.Generics` import, I believe.

**Comment by emilypi** [comment]
@treeowl no worries. I'm surprised hlint didn't catch that on mine. All fixed now, thanks.

**Comment by phadej** [comment]
@emilypi looks good. If you mind squash when you are done polishing, we'll merge.

**Comment by emilypi** [comment]
@phadej beat you to the squash 😄

**Comment by treeowl** [comment]
I'm not very convinced by that `Biff` instance. Is it more likely to be used on purpose or by mistake,

**Comment by emilypi** [comment]
I imagine you'd have to go out of your way to provide `f ~ g => Biff p f g` on purpose. It'd be terribly difficult to it by mistake. Do you have some place we can discuss off line?

**Comment by treeowl** [comment]
I don't immediately see how to prove it's lawful. Can you demonstrate that?

**Comment by phadej** [comment]
With `f ~ g`, `Biff p f g a b` is

```haskell
newtype Biff p f g a b = Biff (p (f a) (g b))
  =< f ~ g >

newype Biff p f f a b = Biff (p (f a) (f b))
```

looks quite symmetric to me. I'm sure you can prove the laws,
using `Swapped p` e.g. `swapped . swapped = id` is quite trivial already.

File: src/Control/Lens/Iso.hs [hunk 2]
```diff
@@ -409,6 +415,57 @@ instance Swapped (,) where
...
+
+instance Swapped ((,,) x) where
+  swapped = iso f f
+    where
+      f ~(x,a,b) = (x,b,a)
```
Comment (glguy) [inline src/Control/Lens/Iso.hs]: Why are these tuple instances using an irrefutable match? The instance for 2-tuples doesn't nor do any of the other new instances for product-types. If there's a good reason can we document it on the instances so people know to rely on it. If there isn't can we switch it to a normal pattern?

Comment (emilypi) [reply src/Control/Lens/Iso.hs]: I'm a proponent of lazy pattern matches for composability reasons, but I'm not tied to them. I figured it's a Kmett library who enjoys going full boar on these things. Considering `swapped` is a good candidate for lazy behavior, it seemed like a good idea.

Comment (glguy) [reply src/Control/Lens/Iso.hs]: OK, then let's make them consistent and document them in the haddock comments

**Comment by phadej** [comment]
Good point about `_1'`. Would it make sense to add strict `swapped'` (I'm not sure whether it can be external/not-class-member combinator).

**Comment by emilypi** [comment]
@phadej I would assume that would be its own combinator in the same way that `_1'` is. I wouldn't expect my `swapped` behavior to be strict by default, which is exacly how this is also how it's done in `Tuple.hs`: https://github.com/ekmett/lens/blob/master/src/Control/Lens/Tuple.hs#L104

**Comment by phadej** [comment]
I meant that i’m not sure that swapped' can be implemented using swapped, or if it have to be class member

**Comment by emilypi** [comment]
Ah, no. I'm not sure if it can.

**Comment by phadej** [comment]
Hmm, I think

```haskell
swapped' = iso f g where
  f x = x `seq` view swapped x
  g y = y `seq` review swapped y
{-# INLINE swapped' #-}
```

would be good enough.

**Comment by treeowl** [comment]
Personally, I tend to be wary of hidden laziness, since it can cause space leaks and slow down code. The lazy instances also aren't strictly law-abiding, because swapping twice may give a more-defined value than doing nothing. But I recognize that Ed has a somewhat different perspective WRT class instances than I do. And the double-swap law *could* be changed to an inequality.

**Comment by phadej** [comment]
@treeowl can lazy `swapped` be implemented in terms of strict `swapped'`?

**Comment by treeowl** [comment]
@phadej, it can with a `Generic` constraint and something confining it to be a record. Arguably, strict swap should only be expressible from lazy swap with an additional `Seqable` (or whatever) constraint; the fact that `seq` is fully polymorphic in Haskell is a bit ugly. In the case of the `Arrow` instance for functions and (therefore?) the `Bifunctor` instance for pairs, `first` and `second` are lazy to support `ArrowLoop`. Is there a major use-case for laziness here?

**Comment by ekmett** [comment]
Merging the PR, any remaining wibbles can be fixed up in-repo.

**Comment by phadej** [comment]
@ekmett your opinion, whether `swapped` should use strict or irrefutable pattern matches?

**Pull request merged**